### ISSUE TITLE
None of the symbols in my package seem to be recognized

### ISSUE THREAD

**Issue by PetrKryslUCSD** [julia-vscode/julia-vscode#800]
I get hundreds of "Missing reference":
![image](https://user-images.githubusercontent.com/9855832/61576415-1a290880-aada-11e9-99e5-9aa10d60b8cd.png)
Yet, running the file `playground` works just fine.

**Comment by davidanthoff** [comment]
Which version of the extension? Can you describe the julia environment situation a bit more in which this is running?

**Comment by PetrKryslUCSD** [comment]
RC 7, on Win 10, with Julia 1.1.1. No documentation is available for anything but standard Julia lib files.

**Comment by davidanthoff** [comment]
What does the VS Code window show as the active environment in the status bar?

**Comment by PetrKryslUCSD** [comment]
![image](https://user-images.githubusercontent.com/9855832/61587175-2fa73c80-ab85-11e9-9ada-750a8f9c4410.png)
I believe it correctly states the environment, and when I point to a symbol initially it shows "Loading" and then there is a long string of error messages (as shown above).

**Comment by eascarrunz** [comment]
I have the exact same problem with Julia 1.1.0 and Julia-VSCode 0.12.0-rc.7 (VSCode 1.36.1)

**Comment by ZacLN** [comment]
This is due to the use of the pattern `using FinEtools.SubModule: something` from within the definition of `FinEtools`. If you try to use the same pattern when including a file from the REPL it'll error, e.g.: 
```julia
module MyModule
module InnerModule
x = 1
end
using MyModule.InnerModule: x
end
```
If you set the environment to that which FinEtools (i.e. Julia 1.x) is installed then things should work (with the slightly odd effect that those submodule imports refer to the variables of the cached module rather than the source files your editing).  Alternatively (and I'd say preferably)  you could import your submodules using the more widespread format of `using .FTypesModule` etc

**Comment by PetrKryslUCSD** [comment]
This is interesting. Replacing `using FinEtools.XYZ` with `using .XYZ` in all the source files of the package works to some degree: in the source files of `FinEtools` itself the documentation look up works. However, doing something like
```
module surfacenormaltest3
using FinEtools
using Test
function test()
   XYZ = reshape([0.0, 0.0], 2, 1)
   tangents = reshape([-1.0, 1.0], 2, 1)
   fe_label = 0
   fi = SurfaceNormal(2)
   @show v = updatenormal!(fi, XYZ::FFltMat, tangents::FFltMat, fe_label::FInt)
   @test v == [0.7071067811865475, 0.7071067811865475]
end
end
using .surfacenormaltest3
surfacenormaltest3.test()
```
where `SurfaceNormal` and `updatenormal!` are defined/exported in/from FinEtools, no documentation is available again.

**Comment by singularitti** [comment]
> This is interesting. Replacing `using FinEtools.XYZ` with `using ..XYZ` in all the source files of the package works to some degree: in the source files of `FinEtools` itself the documentation look up works. However, doing something like
> 
> ```
> module surfacenormaltest3
> using FinEtools
> using Test
> function test()
>    XYZ = reshape([0.0, 0.0], 2, 1)
>    tangents = reshape([-1.0, 1.0], 2, 1)
>    fe_label = 0
>    fi = SurfaceNormal(2)
>    @show v = updatenormal!(fi, XYZ::FFltMat, tangents::FFltMat, fe_label::FInt)
>    @test v == [0.7071067811865475, 0.7071067811865475]
> end
> end
> using .surfacenormaltest3
> surfacenormaltest3.test()
> ```
> 
> where `SurfaceNormal` and `updatenormal!` are defined/exported in/from FinEtools, no documentation is available again.

Me too. Things defined/exported in the top-level module, e.g. [`EquationsOfState`](https://github.com/MineralsCloud/EquationsOfState.jl) is still not recognized.
<img width="605" alt="image" src="https://user-images.githubusercontent.com/25192197/64918193-095ae300-d769-11e9-930c-28c1bb324175.png">

Things defined in its submodules are solved by `using ..XXX`, etc.

Please fix this bug!

**Comment by bilderbuchi** [comment]
Note that in the [OP](https://github.com/julia-vscode/julia-vscode/issues/800#issue-470651219), even `Test` is not recognized. I get the same problem of "Missing reference: Test" with a Julia file that just contains
```
using Test
```
This is on Julia 1.2, Julia plugin 0.12.3.

**Comment by norru** [comment]
Just got this bug. Juila 1.3.0, VSC 1.41.1

**Comment by davidanthoff** [comment]
Also see https://github.com/julia-vscode/julia-vscode/issues/867 with more examples of the same problem.

**Comment by davidanthoff** [comment]
And I'm moving this to the v0.14.1 milestone. We are actively trying fixes for this right now, but they are a bit involved and we want to get 0.14.0 first.

**Comment by davidanthoff** [comment]
https://github.com/julia-vscode/LanguageServer.jl/pull/498 is one attempt to solve this.

**Comment by NLaws** [comment]
Is there a way to disable these false "Missing reference" problems? I have disabled the `Missingrefs` option and restarted VScode and I still get every variable in a JuMP modeled linted.

<img width="288" alt="Screen Shot 2020-04-10 at 7 49 29 AM" src="https://user-images.githubusercontent.com/14351227/78995585-ebd3d900-7aff-11ea-81f7-5c6d8982c265.png">

**Comment by hsm207** [comment]
I have similar problems too and would like to share additional info.

I have this script:

```julia
using Random
using Distributions

Normal()
```

Running in locally, I get a missing reference on `Normal()`.

Running it inside a Docker container, I get a missing reference on both `Random` and `Normal()`.

My system information:

Local Julia:

```
Julia Version 1.4.0
Commit b8e9a9ecc6 (2020-03-21 16:36 UTC)
Platform Info:
  OS: Windows (x86_64-w64-mingw32)
  CPU: Intel(R) Core(TM) i7-7500U CPU @ 2.70GHz
  WORD_SIZE: 64
  LIBM: libopenlibm
  LLVM: libLLVM-8.0.1 (ORCJIT, skylake)
```
Julia inside docker container:

```
Julia Version 1.4.0
Commit b8e9a9ecc6 (2020-03-21 16:36 UTC)
Platform Info:
  OS: Linux (x86_64-pc-linux-gnu)
  CPU: Intel(R) Core(TM) i7-7500U CPU @ 2.70GHz
  WORD_SIZE: 64
  LIBM: libopenlibm
  LLVM: libLLVM-8.0.1 (ORCJIT, skylake)
```

VS Code:

```
Version: 1.45.0-insider (user setup)
Commit: 44c5185373d5f209cd6c0b5d29a216fb801da34e
Date: 2020-04-10T18:12:33.780Z
Electron: 7.2.1
Chrome: 78.0.3904.130
Node.js: 12.8.1
V8: 7.8.279.23-electron.0
OS: Windows_NT x64 10.0.19041
```

Julia extension:

`julialang.language-julia@0.15.19`

**Comment by CameronBieganek** [comment]
It's not clear to me how many of these missing references issues are actual bugs and how many are due to not having environments set up quite right. Maybe what's needed is some documentation on how to have your environments and project directories set up to get correct linting?

One thing I think I've noticed is that if I have a project environment activated and I add some code with a new dependency, e.g.

```julia
using DataFrames

# stuff with data frames
```

I'll get missing reference errors. That's expected and normal. However, if I do `] add DataFrames`, the missing references don't go away until I close and reopen VS Code. It would be nice if the linter immediately picked up changes to the `Project.toml` in the currently active session/environment.

**Comment by davidanthoff** [comment]
Yes, we need to reread project files when they are updated, that is a known problem.

**Comment by tsela** [comment]
I have the same problem with Julia 1.4.2 and the Julia plugin 0.15.40, for a very simply package created using `Pkg.generate()`. The package is called `ProjectSim`, has a correct `Project.toml`, and a `src/ProjectSim.jl` file that defines the module `ProjectSim`. VSCode recognises the environment just fine, I get `Julia env: ProjectSim` in the status bar. Running `using ProjectSim` in the REPL works just fine. But typing `using ProjectSim` in another Julia file anywhere in the project folder (even in `src/`) results in squiggly lines under ProjectSim and no Intellisense working.

**Comment by CameronBieganek** [comment]
@tsela If I understand your example correctly, that's the expected behavior. You can't load a package inside of itself. That would result in infinite recursion. It would be akin to doing this:

```julia
julia> module A
       using A
       end
ERROR: ArgumentError: Package A not found in current path:
- Run `import Pkg; Pkg.add("A")` to install the A package.

Stacktrace:
 [1] require(::Module, ::Symbol) at ./loading.jl:892
```

which throws an error.

**Comment by tsela** [comment]
@CameronBieganek Eh no. I'll be the first to admit I'm a beginner Julia user, but I'm pretty sure that's not what I'm doing. I'm not trying to use a package within itself, I'm trying to use it within its own environment. That's not quite the same thing and is actually completely licit in Julia. I can illustrate it. See the screenshot below:
![Screenshot from 2020-06-06 15-54-30](https://user-images.githubusercontent.com/475638/83946065-ed273680-a80e-11ea-82d8-47be878f72f9.png)
As you can see, I've got a simple `app.jl` script in the root folder of the `ProjectSim` package, which does `using ProjectSim` and creates and prints a simple object of type `Project` (which is defined within the `ProjectSim` module). If I run that script using `julia --project app.jl` in the terminal, it works perfectly fine (see red arrow). As expected, a package's main module is visible in the scope of its own environment. However, as you can also see (yellow arrow), VSCode does not recognise the `ProjectSim` module (see the squiggly lines), despite the Language Server saying it's working in the `ProjectSim` environment (see magenta arrow).
That's the bug: `ProjectSim` is in the scope of its own environment, as `julia --project` proves. So I would expect the Julia Language Server to index it just like it does every other package in that scope (notice how the Dash package is correctly indexed. It's only the local package which isn't).

**Comment by CameronBieganek** [comment]
Ok, that makes sense. Thanks for the additional info. It was the "even in `src/`" part of your first comment that confused me, since most files in `src` are usually `include`-ed inside the package module defined in `src/ProjectSim`.

**Comment by tsela** [comment]
Yeah, I understand, and sorry for being unclear. I just meant that it didn't matter where in the `ProjectSim` folder I put the `app.jl` file, I got the same issue :smiley:.

In the meantime, I found a "workaround" (in quotes because it's very impractical): if I add `include("src/ProjectSim.jl")` before the `using ProjectSim` line in the `app.jl` file and restart the Language Server, it finds the `ProjectSim` module and all the symbols defined therein just fine and Intellisense works. The only problem is that I cannot run the file like that as it gives me an `ERROR: LoadError: importing ProjectSim into Main conflicts with an existing identifier`, which is a perfectly understandable error given what I'm doing. So I need to comment the include whenever I want to run the script, and uncomment it and restart the Language Server whenever I want to actually work on the script. So, a workaround, but an impractical one. Hopefully this problem can be solved soon so I can get rid of that ugly workaround :+1:.

**Comment by Roger-luo** [comment]
I think I have the same problem with @tsela now on v0.16.7, all other modules are correctly indexed, except the project module itself.

I think a better workaround on this is to use a shared environment for the development of the project. The linter seems to be able to detect all the modules in a shared environment.

**Comment by tsela** [comment]
Shared environments cause lots of problems with incompatible packages and are the reason why environments were developed in the first place. I don't think working with a shared environment is a good workaround at all. It's certainly not how you're supposed to work in Julia.

Right now, I use the following workaround that I took from another issue thread about the same problem (but can't find it right now so can't link to it). In order to get the project module itself correctly indexed without having to change anything when running the code, I replace `using ProjectModule` with:

```julia
if isdefined(@__MODULE__, :LanguageServer)
    include("src/ProjectModule.jl")
    using .ProjectModule
else
    using ProjectModule
end
```

It works fine, although a bit ugly to change one's code just because the tooling's indexing behaviour does not match what Julia itself does.

I mean, don't take me wrong, I think the Julia VSCode plugin is fantastic, but Intellisense is supposed to be one of the main strengths of VSCode (and it's certainly the feature people notice first and use/interact with most when using VSCode), and right now it's pretty much broken for Julia. This makes the whole plugin look bad despite the great work the team has put into this.

**Comment by Roger-luo** [comment]
> Shared environments cause lots of problems with incompatible packages and are the reason why environments were developed in the first place. 

@tsela you could create a new shared environment that contains the packages with specific versions you want to develop only. I know this is not as convenient as a local project environment, but at least this is won't require any change of users' code and works better if you have multiple packages under development together. And it is just not practical to change the package code for this and make it into a release, it causes trouble for other developers using other editors.

> I don't think working with a shared environment is a good workaround at all. It's certainly not how you're supposed to work in Julia.

It is just one of many options. You can't develop multiple packages at once with a single local project environment without using `dev`.

Anyway, I'm just listing possible workarounds here. Pick whatever you like. But still hope this issue gets fixed at some point.

**Comment by Roger-luo** [comment]
And I think create a new (fake) shared environment is exactly what julia-vscode/LanguageServer.jl#498 is trying to do (automatically). The problem seems to be that the linter won't scan the local package module for all environments, it only scan what's inside `[deps]`.

**Comment by CameronBieganek** [comment]
Bump. I have a package with a directory structure like this:

```
├── Manifest.toml
├── Project.toml
├── README.md
├── scripts
│   ├── get_anchor_xml.jl
│   ├── get_ptsd_citations_distribution.jl
│   ├── run_scopus_citation_graph.jl
│   └── test_scopus_query.jl
├── src
│   ├── CitationGraphs.jl
│   ├── citation_graph.jl
│   ├── metrics.jl
│   ├── pubmed_query.jl
│   └── scopus_query.jl
└── test
    ├── Manifest.toml
    ├── Project.toml
    └── runtests.jl
```

All of the scripts in the `scripts` folder are riddled with `Missing reference` linter warnings. It makes my scripts nearly illegible. I see that the milestone for this issue is "Backlog". In my opinion, this issue should be a higher priority, since it makes entire files unreadable.

**Comment by aviatesk** [comment]
use `Julia > Lint > Disabled Dirs` configuration for the time being.

**Comment by davidanthoff** [comment]
We're pretty actively working on fixing all of these things, but it is also a fairly long road...

**Comment by Roger-luo** [comment]
I'm wondering if there is any progress/updates on this one year later? or is there a roadmap of what a proper solution to this problem can be implemented?

**Comment by XilinJia** [comment]
I've just come back to Julia after 3 years absence, not quite update on how this goes.  Seems like I am having the same issue with Julia extenstion v1.5.11 on VSCode.

**Comment by PetrKryslUCSD** [comment]
See https://discourse.julialang.org/t/vs-code-no-implementation-found-in-developed-package/77198/2
I can see the problem with tracking evolving code, but still, there may be some clever solution.

**Comment by kcajf** [comment]
> I've just come back to Julia after 3 years absence, not quite update on how this goes.  Seems like I am having the same issue with Julia extenstion v1.5.11 on VSCode.

It's still a problem for me - I'm unaware of a proper solution.

**Comment by charleskawczynski** [comment]
Would this issue explain why goto definitions sometimes don't work?

**Comment by davidanthoff** [comment]
The language server backend was rewritten in v1.220.0: StaticLint.jl and SymbolServer.jl were folded into JuliaWorkspaces, and completion, hover, signature help, go-to-definition and find-references were reimplemented. This report predates that rewrite, so the code that produced it no longer exists and the behaviour cannot be confirmed against the current engine.

Closing as outdated. If you still see this on a current release, please open a fresh issue with a reproducer and we will look at it against the new engine.
